README.md
# dwebstore
[![Build Status](https://travis-ci.com/andrewosh/dwebstore.svg?token=WgJmQm3Kc6qzq1pzYrkx&branch=master)](https://travis-ci.com/andrewosh/dwebstore)

This module is the canonical implementation of the "dwebstore" interface, which exposes a DDatabase factory and a set of associated functions for managing generated Hypercores.

A dwebstore is designed to efficiently store and replicate multiple sets of interlinked Hypercores, such as those used by [DDrive](https://github.com/distributedweb/ddrive) and [mountable-dwebtrie](https://github.com/andrewosh/dwebtrie), removing the responsibility of managing custom storage/replication code from these higher-level modules.

In order to do this, dwebstore provides:
1. __Key derivation__ - all writable DDatabase keys are derived from a single master key.
2. __Caching__ - Two separate caches are used for passively replicating cores (those requested by peers) and active cores (those requested by the owner of the dwebstore).
3. __Storage bootstrapping__ - You can create a `default` DDatabase that will be loaded when a key is not specified, which is useful when you don't want to reload a previously-created DDatabase by key.
4. __Namespacing__ - If you want to create multiple compound data structures backed by a single dwebstore, you can create namespaced corestores such that each data structure's `default` feed is separate.

### Installation
`npm i dwebstore --save`

### Usage
A dwebstore instance can be constructed with a random-access-storage module, a function that returns a random-access-storage module given a path, or a string. If a string is specified, it will be assumed to be a path to a local storage directory:
```js
const DWebstore = require('dwebstore')
const ram = require('random-access-memory')
const store = new DWebstore(ram)
await store.ready()
```

Hypercores can be generated with both the `get` and `default` methods. If the first writable core is created with `default`, it will be used for storage bootstrapping. We can always reload this bootstrapping core off disk without your having to store its public key externally. Keys for other hypercores should either be stored externally, or referenced from within the default core:
```js
const core1 = store1.default()
```
_Note: You do not have to create a default feed before creating additional ones unless you'd like to bootstrap your dwebstore from disk the next time it's instantiated._

Additional hypercores can be created by key, using the `get` method. In most scenarios, these additional keys can be extracted from the default (bootstrapping) core. If that's not the case, keys will have to be stored externally:
```js
const core2 = store1.get({ key: Buffer(...) })
```
All hypercores are indexed by their discovery keys, so that they can be dynamically injected into replication streams when requested.

Two corestores can be replicated with the `replicate` function, which accepts ddatabase's `replicate` options:
```js
const store1 = new DWebstore(ram)
const store2 = new DWebstore(ram)
await Promise.all([store1.ready(), store2.ready()]

const core1 = store2.get()
const core2 = store2.get({ key: core1.key })
const stream = store1.replicate(true, { live: true })
stream.pipe(store2.replicate(false, { live: true })).pipe(stream) // This will replicate all common cores.
```

### API
#### `const store = dwebstore(storage, [opts])`
Create a new dwebstore instance. `storage` can be either a random-access-storage module, or a function that takes a path and returns a random-access-storage instance.

Opts is an optional object which can contain any DDatabase constructor options, plus the following:
```js
{
  cacheSize: 1000 // The size of the LRU cache for passively-replicating cores.
}
```

#### `store.default(opts)`
Create a new default ddatabase, which is used for bootstrapping the creation of subsequent hypercores. Options match those in `get`.

#### `store.get(opts)`
Create a new ddatabase. Options can be one of the following:
```js
{
  key: 0x1232..., // A Buffer representing a ddatabase key
  discoveryKey: 0x1232..., // A Buffer representing a ddatabase discovery key (must have been previously created by key)
  ...opts // All other options accepted by the ddatabase constructor
}
```

If `opts` is a Buffer, it will be interpreted as a ddatabase key.

#### `store.on('feed', feed, options)`

Emitted everytime a feed is loaded internally (ie, the first time get(key) is called).
Options will be the full options map passed to .get.

#### `store.replicate(isInitiator, [opts])`
Create a replication stream that will replicate all cores currently in memory in the dwebstore instance.

When piped to another dwebstore's replication stream, only those cores that are shared between the two corestores will be successfully replicated.

#### `store.list()`
Returns a Map of all cores currently cached in memory. For each core in memory, the map will contain the following entries:
```
{
  discoveryKey => core,
  ...
}
```

#### `const namespacedStore = store.namespace('some-name')`
Create a "namespaced" dwebstore that uses the same underlying storage as its parent, and mirrors the complete dwebstore API. 

`namespacedStore.default` returns a different default core, using the namespace as part of key generation, which makes it easier to bootstrap multiple data structures from the same dwebstore. The general pattern is for all data structures to bootstrap themselves from their dwebstore's default feed:
```js
const store = new DWebstore(ram)
const drive1 = new DDrive(store.namespace('drive1'))
const drive2 = new DDrive(store.namespace('drive2'))
```

Namespaces currently need to be saved separately outside of dwebstore (as a mapping from key to namespace), so that data structures remain writable across restarts. Extending the above code, this might look like:
```js
async function getDrive (opts = {}) {
  let namespace = opts.key ? await lookupNamespace(opts.key) : await createNamespace()
  const namespacedCorestore = store.namespace(namespace)
  const drive = new DDrive(namespacedCorestore)
  await saveNamespace(drive.key, namespace)
}
```

#### `store.close(cb)`
Close all hypercores previously generated by the dwebstore.

### License
MIT


CHANGELOG.md
## 5.7.0
- Simplifies internal reference counting and namespace handling. This removes the `NamespacedCorestore` class, but does not alter the interface.
- Uses `refpool` for reference handling.
- Removes `Nanoguard` and the undocumented `this.guard` property on DWebstore.
- Removes the private `_name` option to `DWebstore.get` in favor of a public `name` option.


test/all.js
const p = require('path')
const ram = require('random-access-memory')
const raf = require('random-access-file')
const datEncoding = require('dwebx-encoding')
const hypercoreCrypto = require('ddatabase-crypto')
const test = require('tape')

const DWebstore = require('..')
const {
  runAll,
  validateCore,
  cleanup
} = require('./helpers')

test('ram-based dwebstore, different get options', async t => {
  const store1 = await create(ram)
  const core1 = store1.default()
  var core2, core3, core4, core5, core6

  await runAll([
    cb => core1.ready(cb),
    cb => core1.append('hello', cb),
    cb => {
      // Buffer arg
      core2 = store1.get(core1.key)
      return core2.ready(cb)
    },
    cb => {
      // Object arg
      core3 = store1.get({ key: core1.key })
      return core3.ready(cb)
    },
    cb => {
      // Discovery key option
      core4 = store1.get({ discoveryKey: core1.discoveryKey })
      return core4.ready(cb)
    },
    cb => {
      // String option
      core5 = store1.get({ key: datEncoding.encode(core1.key) })
      return core5.ready(cb)
    },
    cb => {
      // Custom keypair option
      core6 = store1.get({ keyPair: { secretKey: core1.secretKey, publicKey: core1.key } })
      return core6.ready(cb)
    }
  ])

  t.same(core1, core2)
  t.same(core1, core3)
  t.same(core1, core4)
  t.same(core1, core5)
  t.same(core1, core6)
  t.end()
})

test('ram-based dwebstore, simple replication', async t => {
  const store1 = await create(ram)
  const store2 = await create(ram)
  const core1 = store1.default()
  const core2 = store1.get()
  var core3 = null
  var core4 = null

  await runAll([
    cb => core1.ready(cb),
    cb => core2.ready(cb),
    cb => {
      core3 = store2.default(core1.key)
      return core3.ready(cb)
    },
    cb => {
      core4 = store2.get({ key: core2.key })
      return core4.ready(cb)
    },
    cb => core1.append('hello', cb),
    cb => core1.append('world', cb),
    cb => core2.append('cat', cb),
    cb => core2.append('dog', cb),
    cb => {
      const stream = store1.replicate(true)
      stream.pipe(store2.replicate(false)).pipe(stream)
      stream.on('end', cb)
    }
  ])

  await validateCore(t, core3, [Buffer.from('hello'), Buffer.from('world')])
  await validateCore(t, core4, [Buffer.from('cat'), Buffer.from('dog')])

  t.end()
})

test('ram-based dwebstore, replicating with different default keys', async t => {
  const store1 = await create(ram)
  const store2 = await create(ram)
  const core1 = store1.default()
  const core2 = store1.get()
  var core3 = null
  var core4 = null

  await runAll([
    cb => core1.ready(cb),
    cb => core2.ready(cb),
    cb => {
      core3 = store2.default()
      return core3.ready(cb)
    },
    cb => {
      core4 = store2.get({ key: core1.key })
      return core4.ready(cb)
    },
    cb => core1.append('cat', cb),
    cb => core1.append('dog', cb),
    cb => {
      const stream = store1.replicate(true)
      stream.pipe(store2.replicate(false)).pipe(stream)
      stream.on('end', cb)
    }
  ])

  await validateCore(t, core4, [Buffer.from('cat'), Buffer.from('dog')])
  t.end()
})

test('ram-based dwebstore, sparse replication', async t => {
  const store1 = await create(ram, { sparse: true })
  const store2 = await create(ram, { sparse: true })
  const core1 = store1.default()
  const core2 = store1.get()
  var core3 = null
  var core4 = null

  await runAll([
    cb => core1.ready(cb),
    cb => core2.ready(cb),
    cb => {
      t.same(core2.sparse, true)
      t.same(core1.sparse, true)
      return process.nextTick(cb, null)
    },
    cb => {
      core3 = store2.default(core1.key)
      return core3.ready(cb)
    },
    cb => {
      core4 = store2.get({ key: core2.key })
      return core4.ready(cb)
    },
    cb => {
      const stream = store1.replicate(true, { live: true })
      stream.pipe(store2.replicate(false, { live: true })).pipe(stream)
      return process.nextTick(cb, null)
    },
    cb => core1.append('hello', cb),
    cb => core1.append('world', cb),
    cb => core2.append('cat', cb),
    cb => core2.append('dog', cb),
    cb => {
      t.same(core3.length, 0)
      t.same(core4.length, 0)
      return process.nextTick(cb, null)
    }
  ])

  await validateCore(t, core3, [Buffer.from('hello'), Buffer.from('world')])
  await validateCore(t, core4, [Buffer.from('cat'), Buffer.from('dog')])
  t.end()
})

test('ram-based dwebstore, sparse replication with different default keys', async t => {
  const store1 = await create(ram, { sparse: true })
  const store2 = await create(ram, { sparse: true })
  const core1 = store1.default()
  var core3 = null
  var core4 = null

  await runAll([
    cb => core1.ready(cb),
    cb => {
      core3 = store2.default()
      return core3.ready(cb)
    },
    cb => {
      const s1 = store1.replicate(true, { live: true })
      const s2 = store2.replicate(false, { live: true })
      s1.pipe(s2).pipe(s1)
      return process.nextTick(cb, null)
    },
    cb => core1.append('cat', cb),
    cb => core1.append('dog', cb),
    cb => {
      core4 = store2.get({ key: core1.key })
      return core4.ready(cb)
    },
    cb => {
      t.same(core4.length, 0)
      t.same(core1.length, 2)
      return process.nextTick(cb, null)
    }
  ])

  await validateCore(t, core4, [Buffer.from('cat'), Buffer.from('dog')])
  t.end()
})

test('raf-based dwebstore, simple replication', async t => {
  const store1 = await create(path => raf(p.join('store1', path)))
  const store2 = await create(path => raf(p.join('store2', path)))
  const core1 = store1.default()
  const core2 = store1.get()
  var core3 = null
  var core4 = null

  await runAll([
    cb => core1.ready(cb),
    cb => core2.ready(cb),
    cb => {
      core3 = store2.default({ key: core1.key })
      return core3.ready(cb)
    },
    cb => {
      core4 = store2.get({ key: core2.key })
      return core4.ready(cb)
    },
    cb => core1.append('hello', cb),
    cb => core1.append('world', cb),
    cb => core2.append('cat', cb),
    cb => core2.append('dog', cb),
    cb => {
      setImmediate(() => {
        const stream = store1.replicate(true)
        stream.pipe(store2.replicate(false)).pipe(stream)
        stream.on('end', cb)
      })
    }
  ])

  await validateCore(t, core3, [Buffer.from('hello'), Buffer.from('world')])
  await validateCore(t, core4, [Buffer.from('cat'), Buffer.from('dog')])
  await cleanup(['store1', 'store2'])
  t.end()
})

test('raf-based dwebstore, close and reopen', async t => {
  var store = await create('test-store')
  var firstCore = store.default()
  var reopenedCore = null

  await runAll([
    cb => firstCore.ready(cb),
    cb => firstCore.append('hello', cb),
    cb => store.close(cb),
    cb => {
      t.true(firstCore.closed)
      return process.nextTick(cb, null)
    },
    cb => {
      create('test-store').then(store => {
        reopenedCore = store.default()
        return reopenedCore.ready(cb)
      })
    }
  ])

  await validateCore(t, reopenedCore, [Buffer.from('hello')])
  await cleanup(['test-store'])
  t.end()
})

test('raf-based dwebstore, close and reopen with keypair option', async t => {
  var store = await create('test-store')
  const keyPair = hypercoreCrypto.keyPair()
  var firstCore = store.get({ keyPair })
  var reopenedCore = null

  await runAll([
    cb => firstCore.ready(cb),
    cb => firstCore.append('hello', cb),
    cb => store.close(cb),
    cb => {
      t.true(firstCore.closed)
      return process.nextTick(cb, null)
    },
    cb => {
      create('test-store').then(store => {
        reopenedCore = store.get({ key: keyPair.publicKey })
        return reopenedCore.ready(cb)
      })
    }
  ])

  await validateCore(t, reopenedCore, [Buffer.from('hello')])
  t.true(reopenedCore.writable)

  await cleanup(['test-store'])
  t.end()
})

test('live replication with an additional core', async t => {
  const store1 = await create(ram)
  const store2 = await create(ram)

  const core1 = store1.default()
  var core2 = null
  var core3 = null
  var core4 = null

  await runAll([
    cb => core1.ready(cb),
    cb => {
      core3 = store2.default({ key: core1.key })
      return core3.ready(cb)
    },
    cb => {
      const stream = store1.replicate(true, { live: true })
      stream.pipe(store2.replicate(false, { live: true })).pipe(stream)
      return cb(null)
    },
    cb => {
      core2 = store1.get()
      return core2.ready(cb)
    },
    cb => {
      core4 = store2.get(core2.key)
      return core4.ready(cb)
    },
    cb => core2.append('hello', cb),
    cb => core2.append('world', cb)
  ])

  await validateCore(t, core4, [Buffer.from('hello'), Buffer.from('world')])
  t.end()
})

test('namespaced corestores use separate default keys', async t => {
  const store1 = await create(ram)
  const store2 = store1.namespace('store2')
  const store3 = store1.namespace('store3')

  await store2.ready()
  await store3.ready()

  const feed1 = store2.default()
  const feed2 = store3.default()

  t.true(!feed1.key.equals(feed2.key))

  t.end()
})

test('namespaced corestores will not increment reference multiple times', async t => {
  const store1 = await create(ram)
  const store2 = store1.namespace('store2')
  const store3 = store1.namespace('store3')

  await store2.ready()
  await store3.ready()

  const feed1 = store2.default()
  await feed1.ready()
  const feed3 = store3.get({ key: feed1.key })
  const feed4 = store3.get({ key: feed1.key })
  const feed5 = store3.get({ key: feed1.key })

  t.same(feed1, feed3)
  t.same(feed1, feed4)
  t.same(feed1, feed5)

  const entry = store1.cache.entry(feed1.discoveryKey.toString('hex'))
  t.same(entry.refs, 2)

  t.end()
})

test('namespaced corestores can be nested', async t => {
  const store1 = await create(ram)
  const store2 = store1.namespace('store2')
  const store1a = store1.namespace('a')
  const store2a = store2.namespace('a')

  const feed1 = store1.default()
  const feed2 = store2.default()
  const feed1a = store1a.default()
  const feed2a = store2a.default()

  await feed1.ready()
  await feed2.ready()
  await feed1a.ready()
  await feed2a.ready()

  t.notEqual(feed1a.key, feed2a.key)

  t.end()
})

test('caching works correctly when reopening by discovery key', async t => {
  var store = await create('test-store')
  var firstCore = store.default()
  var discoveryKey = null
  var reopenedCore = null

  await runAll([
    cb => firstCore.ready(cb),
    cb => {
      discoveryKey = firstCore.discoveryKey
      return cb(null)
    },
    cb => firstCore.append('hello', cb),
    cb => store.close(cb),
    cb => {
      t.true(firstCore.closed)
      return process.nextTick(cb, null)
    },
    cb => {
      create('test-store').then(reopenedStore => {
        store = reopenedStore
        reopenedCore = store.get({ discoveryKey })
        return reopenedCore.ready(cb)
      })
    },
    cb => {
      const idx = discoveryKey.toString('hex')
      t.true(store.cache.has(idx))
      return cb(null)
    }
  ])

  await validateCore(t, reopenedCore, [Buffer.from('hello')])
  await cleanup(['test-store'])
  t.end()
})

test('can check if cores are loaded', async t => {
  const store = await create(ram)
  await store.ready()

  const feed1 = store.default()
  const feed2 = store.get()
  const badKey = hypercoreCrypto.randomBytes(32)
  const badDiscoveryKey = hypercoreCrypto.discoveryKey(badKey)

  t.true(store.isLoaded({ key: feed1.key }))
  t.true(store.isLoaded({ discoveryKey: feed1.discoveryKey }))
  t.true(store.isLoaded({ discoveryKey: feed2.discoveryKey }))
  t.false(store.isLoaded({ key: badKey }))
  t.false(store.isLoaded({ key: badDiscoveryKey }))

  t.end()
})

test('top-level dwebstore replicates all opened cores', async t => {
  const store1 = await create(ram)
  const store2 = await create(ram)

  const core1 = store1.default()
  const ns1 = store1.namespace()
  var core2 = ns1.default()
  var core3 = null

  await runAll([
    cb => core1.ready(cb),
    cb => core2.ready(cb),
    cb => {
      // Only replicate the top-level dwebstore
      const stream = store1.replicate(true, { live: true })
      stream.pipe(store2.replicate(false, { live: true })).pipe(stream)
      return cb(null)
    },
    cb => {
      core3 = store2.get(core2.key)
      return core3.ready(cb)
    },
    cb => core2.append('hello', cb),
    cb => core2.append('world', cb)
  ])

  await validateCore(t, core3, [Buffer.from('hello'), Buffer.from('world')])
  t.end()
})
async function create (storage, opts) {
  const store = new DWebstore(storage, opts)
  await store.ready()
  return store
}


index.js
const HypercoreProtocol = require('ddatabase-protocol')
const Nanoresource = require('nanoresource/emitter')
const ddatabase = require('ddatabase')
const hypercoreCrypto = require('ddatabase-crypto')
const datEncoding = require('dwebx-encoding')
const maybe = require('call-me-maybe')

const RefPool = require('refpool')
const deriveSeed = require('derive-key')
const derivedStorage = require('derived-key-storage')
const raf = require('random-access-file')

const MASTER_KEY_FILENAME = 'master_key'
const NAMESPACE = 'dwebstore'
const NAMESPACE_SEPERATOR = ':'

class InnerCorestore extends Nanoresource {
  constructor (storage, opts = {}) {
    super()

    if (typeof storage === 'string') storage = defaultStorage(storage)
    if (typeof storage !== 'function') throw new Error('Storage should be a function or string')
    this.storage = storage

    this.opts = opts

    this._replicationStreams = []
    this.cache = new RefPool({
      maxSize: opts.cacheSize || 1000,
      close: core => {
        core.close(err => {
          if (err) this.emit('error', err)
        })
      }
    })

    // Generated in _open
    this._masterKey = opts.masterKey || null
    this._id = hypercoreCrypto.randomBytes(8)
  }

  // Nanoresource Methods

  _open (cb) {
    if (this._masterKey) return cb()
    const keyStorage = this.storage(MASTER_KEY_FILENAME)
    keyStorage.read(0, 32, (err, key) => {
      if (err) {
        this._masterKey = hypercoreCrypto.randomBytes(32)
        return keyStorage.write(0, this._masterKey, err => {
          if (err) return cb(err)
          keyStorage.close(cb)
        })
      }
      this._masterKey = key
      keyStorage.close(cb)
    })
  }

  _close (cb) {
    let error = null
    for (const { stream } of this._replicationStreams) {
      stream.destroy()
    }
    if (!this.cache.size) return process.nextTick(cb, null)
    let remaining = this.cache.size
    for (const { value: core } of this.cache.entries.values()) {
      core.close(err => {
        if (err) error = err
        if (!--remaining) {
          if (error) return cb(error)
          return cb(null)
        }
      })
    }
  }

  // Private Methods

  _checkIfExists (dkey, cb) {
    dkey = encodeKey(dkey)
    if (this.cache.has(dkey)) return process.nextTick(cb, null, true)

    const coreStorage = this.storage([dkey.slice(0, 2), dkey.slice(2, 4), dkey, 'key'].join('/'))

    coreStorage.read(0, 32, (err, key) => {
      if (err) return cb(err)
      coreStorage.close(err => {
        if (err) return cb(err)
        if (!key) return cb(null, false)
        return cb(null, true)
      })
    })
  }

  _injectIntoReplicationStreams (core) {
    for (const { stream, opts } of this._replicationStreams) {
      this._replicateCore(false, core, stream, { ...opts })
    }
  }

  _replicateCore (isInitiator, core, mainStream, opts) {
    if (!core) return
    core.ready(function (err) {
      if (err) return
      core.replicate(isInitiator, {
        ...opts,
        stream: mainStream
      })
    })
  }

  _deriveSecret (namespace, name) {
    return deriveSeed(namespace, this._masterKey, name)
  }

  _generateKeyPair (name) {
    if (typeof name === 'string') name = Buffer.from(name)
    else if (!name) name = hypercoreCrypto.randomBytes(32)

    const seed = this._deriveSecret(NAMESPACE, name)

    const keyPair = hypercoreCrypto.keyPair(seed)
    const discoveryKey = hypercoreCrypto.discoveryKey(keyPair.publicKey)
    return { name, publicKey: keyPair.publicKey, secretKey: keyPair.secretKey, discoveryKey }
  }

  _generateKeys (coreOpts) {
    if (Buffer.isBuffer(coreOpts)) coreOpts = { key: coreOpts }

    if (coreOpts.keyPair) {
      const publicKey = coreOpts.keyPair.publicKey
      const secretKey = coreOpts.keyPair.secretKey
      return {
        publicKey,
        secretKey,
        discoveryKey: hypercoreCrypto.discoveryKey(publicKey),
        name: null
      }
    }
    if (coreOpts.key) {
      const publicKey = decodeKey(coreOpts.key)
      return {
        publicKey,
        secretKey: null,
        discoveryKey: hypercoreCrypto.discoveryKey(publicKey),
        name: null
      }
    }
    if (coreOpts.default || coreOpts.name) {
      if (!coreOpts.name) throw new Error('If the default option is set, a name must be specified.')
      return this._generateKeyPair(coreOpts.name)
    }
    if (coreOpts.discoveryKey) {
      const discoveryKey = decodeKey(coreOpts.discoveryKey)
      return {
        publicKey: null,
        secretKey: null,
        discoveryKey,
        name: null
      }
    }
    return this._generateKeyPair(null)
  }

  // Public Methods

  isLoaded (coreOpts) {
    const generatedKeys = this._generateKeys(coreOpts)
    return this.cache.has(encodeKey(generatedKeys.discoveryKey))
  }

  isExternal (coreOpts) {
    const generatedKeys = this._generateKeys(coreOpts)
    const entry = this._cache.entry(encodeKey(generatedKeys.discoveryKey))
    if (!entry) return false
    return entry.refs !== 0
  }

  get (coreOpts = {}) {
    if (!this.opened) throw new Error('DWebstore.ready must be called before get.')
    const self = this

    const generatedKeys = this._generateKeys(coreOpts)
    const { publicKey, discoveryKey, secretKey } = generatedKeys
    const id = encodeKey(discoveryKey)

    const cached = this.cache.get(id)
    if (cached) return cached

    const storageRoot = [id.slice(0, 2), id.slice(2, 4), id].join('/')

    const keyStorage = derivedStorage(createStorage, (name, cb) => {
      if (name) {
        const res = this._generateKeyPair(name)
        if (discoveryKey && (!discoveryKey.equals((res.discoveryKey)))) {
          return cb(new Error('Stored an incorrect name.'))
        }
        return cb(null, res)
      }
      if (secretKey) return cb(null, generatedKeys)
      if (publicKey) return cb(null, { name: null, publicKey, secretKey: null })
      const err = new Error('Unknown key pair.')
      err.unknownKeyPair = true
      return cb(err)
    })

    const cacheOpts = { ...this.opts.cache }
    if (coreOpts.cache) {
      if (coreOpts.cache.data === false) delete cacheOpts.data
      if (coreOpts.cache.tree === false) delete cacheOpts.tree
    }
    if (cacheOpts.data) cacheOpts.data = cacheOpts.data.namespace()
    if (cacheOpts.tree) cacheOpts.tree = cacheOpts.tree.namespace()

    const core = ddatabase(name => {
      if (name === 'key') return keyStorage.key
      if (name === 'secret_key') return keyStorage.secretKey
      return createStorage(name)
    }, publicKey, {
      ...this.opts,
      ...coreOpts,
      cache: cacheOpts,
      createIfMissing: !!publicKey
    })

    this.cache.set(id, core)
    core.ifAvailable.wait()

    var errored = false
    core.once('error', onerror)
    core.once('ready', onready)
    core.once('close', onclose)

    return core

    function onready () {
      if (errored) return
      self.emit('feed', core, coreOpts)
      core.removeListener('error', onerror)
      self._injectIntoReplicationStreams(core)
      // TODO: nexttick here needed? prob not, just legacy
      process.nextTick(() => core.ifAvailable.continue())
    }

    function onerror (err) {
      errored = true
      core.ifAvailable.continue()
      self.cache.delete(id)
      if (err.unknownKeyPair) {
        // If an error occurs during creation by discovery key, then that core does not exist on disk.
        // TODO: This should not throw, but should propagate somehow.
      }
    }

    function onclose () {
      self.cache.delete(id)
    }

    function createStorage (name) {
      return self.storage(storageRoot + '/' + name)
    }
  }

  replicate (isInitiator, cores, replicationOpts = {}) {
    const self = this

    const finalOpts = { ...this.opts, ...replicationOpts }
    const mainStream = replicationOpts.stream || new HypercoreProtocol(isInitiator, { ...finalOpts })
    var closed = false

    for (const core of cores) {
      this._replicateCore(isInitiator, core, mainStream, { ...finalOpts })
    }

    mainStream.on('discovery-key', ondiscoverykey)
    mainStream.on('finish', onclose)
    mainStream.on('end', onclose)
    mainStream.on('close', onclose)

    const streamState = { stream: mainStream, opts: finalOpts }
    this._replicationStreams.push(streamState)

    return mainStream

    function ondiscoverykey (dkey) {
      // Get will automatically add the core to all replication streams.
      self._checkIfExists(dkey, (err, exists) => {
        if (closed) return
        if (err || !exists) return mainStream.close(dkey)
        const passiveCore = self.get({ discoveryKey: dkey })
        self._replicateCore(false, passiveCore, mainStream, { ...finalOpts })
      })
    }

    function onclose () {
      if (!closed) {
        self._replicationStreams.splice(self._replicationStreams.indexOf(streamState), 1)
        closed = true
      }
    }
  }
}

class DWebstore extends Nanoresource {
  constructor (storage, opts = {}) {
    super()

    this.storage = storage
    this.name = opts.name || 'default'
    this.inner = opts.inner || new InnerCorestore(storage, opts)
    this.cache = this.inner.cache
    this.store = this // Backwards-compat for NamespacedCorestore

    this._parent = opts.parent
    this._isNamespaced = !!opts.name
    this._openedCores = new Map()

    const onfeed = feed => this.emit('feed', feed)
    const onerror = err => this.emit('error', err)
    this.inner.on('feed', onfeed)
    this.inner.on('error', onerror)
    this._unlisten = () => {
      this.inner.removeListener('feed', onfeed)
      this.inner.removeListener('error', onerror)
    }
  }

  ready (cb) {
    return maybe(cb, new Promise((resolve, reject) => {
      this.open(err => {
        if (err) return reject(err)
        return resolve()
      })
    }))
  }

  // Nanoresource Methods

  _open (cb) {
    return this.inner.open(cb)
  }

  _close (cb) {
    this._unlisten()
    if (!this._parent) return this.inner.close(cb)
    for (const dkey of this._openedCores) {
      this.cache.decrement(dkey)
    }
    return process.nextTick(cb, null)
  }

  // Private Methods

  _maybeIncrement (core) {
    const id = encodeKey(core.discoveryKey)
    if (this._openedCores.has(id)) return
    this._openedCores.set(id, core)
    this.cache.increment(id)
  }

  // Public Methods

  get (coreOpts = {}) {
    if (Buffer.isBuffer(coreOpts)) coreOpts = { key: coreOpts }
    const core = this.inner.get(coreOpts)
    this._maybeIncrement(core)
    return core
  }

  default (coreOpts = {}) {
    if (Buffer.isBuffer(coreOpts)) coreOpts = { key: coreOpts }
    return this.get({ ...coreOpts, name: this.name })
  }

  namespace (name) {
    if (!name) name = hypercoreCrypto.randomBytes(32)
    if (Buffer.isBuffer(name)) name = name.toString('hex')
    name = this._isNamespaced ? this.name + NAMESPACE_SEPERATOR + name : name
    return new DWebstore(this.storage, {
      inner: this.inner,
      parent: this,
      name
    })
  }

  replicate (isInitiator, opts) {
    const cores = !this._parent ? allReferenced(this.cache) : this._openedCores.values()
    return this.inner.replicate(isInitiator, cores, opts)
  }

  isLoaded (coreOpts) {
    return this.inner.isLoaded(coreOpts)
  }

  isExternal (coreOpts) {
    return this.inner.isExternal(coreOpts)
  }

  list () {
    return new Map([...this._openedCores])
  }
}

function * allReferenced (cache) {
  for (const entry of cache.entries.values()) {
    if (entry.refs > 0) yield entry.value
    continue
  }
}

function encodeKey (key) {
  return Buffer.isBuffer(key) ? datEncoding.encode(key) : key
}

function decodeKey (key) {
  return (typeof key === 'string') ? datEncoding.decode(key) : key
}

function defaultStorage (dir) {
  return function (name) {
    try {
      var lock = name.endsWith('/bitfield') ? require('fd-lock') : null
    } catch (err) {}
    return raf(name, { directory: dir, lock: lock })
  }
}

module.exports = DWebstore
